File: src/Notificator.kt
import java.io.*
import java.util.*
import java.lang.Exception

const val KEY_ICON_PATH = "ICON_PATH"
const val KEY_NOTES_PATH = "NOTES_PATH"
const val KEY_MINIMAL_SPACE = "MINIMAL_SPACE"
const val KEY_ADDITIONAL_SPACE = "ADDITIONAL_SPACE"

// path to default place where xpad is storing data
val DEFAULT_NOTES_PATH = System.getProperty("user.home") + "/.config/xpad" // this can't be const
const val DEFAULT_MINIMAL_SPACE = "5" // todo simplyfi that system
const val DEFAULT_ADDI_SPACE_AMOUNT = "10"

val prop = getConf()
val config = myProp()

data class myProp constructor(
        val notesPath: String,
        val iconPath: String,
        val minimalSpace: Int,
        val additionalSpace: Int
) {

    constructor(): this(prop.getProperty(KEY_NOTES_PATH, DEFAULT_NOTES_PATH),
                        prop.getProperty(KEY_ICON_PATH),
                        Integer.parseInt(prop.getProperty(KEY_MINIMAL_SPACE, DEFAULT_MINIMAL_SPACE)),
                        Integer.parseInt(prop.getProperty(KEY_ADDITIONAL_SPACE, DEFAULT_ADDI_SPACE_AMOUNT)))
}

fun main(args: Array<String>) {

    var wantHelp = false
    var onlyInCommandLine = false

    for (arg in args) {
        if (arg.startsWith("-")) {
            if (arg.contains('c')) {
                onlyInCommandLine = true
            }
            if (arg.contains('h')) {
                wantHelp = true
            }
        }
    }

    var head: String
    var content: Array<String>
    when {          // you can easily declare how flags change behavior of the script inside 'when'
        wantHelp -> {
            //TODO update instructions how to use this script
            head = "Help";
            content = arrayOf("Help",
                    "How to use: ",
                    "Flags:",
                    "    -c show notes only in command line",
                    "    -h show help",
                    "Config:",
                    "   Go to .config/Notificator to edit config file.",
                    "   You can set or change icon of the notification,",
                    "   change path to where xpad is storing notes,",
                    "   minimal length of notification ",
                    "   and how much notification will be elongated.")
        }

        else -> try {
            head = "TODO"
            content = getNotesContent(config.notesPath)
        } catch (exe: Exception) {
            head = "Error :("
            content = arrayOf(exe.message.toString())
        }
    }

    showNotifi(head, content, onlyInCommandLine)
}

private fun getConf(): Properties {
    val prop = Properties()

    // this script is Linux only so using Path builder is useless because we know file that file name separator is '/'
    val confDir = File(System.getProperty("user.home") + "/" + ".config" + "/" + "Notificator")
    val confFile = File(confDir.absolutePath + "/" + "notificator.conf")

    if (confDir.mkdir() || confFile.createNewFile()) {

        with(prop) {
            setProperty(KEY_ICON_PATH, "CHANGE_ME")
            setProperty(KEY_NOTES_PATH, DEFAULT_NOTES_PATH)
            setProperty(KEY_MINIMAL_SPACE, DEFAULT_MINIMAL_SPACE)
            setProperty(KEY_ADDITIONAL_SPACE, DEFAULT_ADDI_SPACE_AMOUNT.length.toString())
        }

        saveProperties(prop, confFile)
    } else {
        loadProperties(prop, confFile)
    }

//    prop.toList().forEach { println(it) } // TESTING
    return prop
}

fun saveProperties(properties: Properties, file: File) {
    val output = FileOutputStream(file)
    properties.store(output, "Properties for Notificator, if you want icon for your notifications pleas change ICON_PATH")
    output.close()
}

fun loadProperties(properties: Properties, file: File) {
    val input = FileInputStream(file)
    properties.load(input)
    input.close()
}

private fun getNotesContent(path: String): Array<String> {

    val listOfFiles = File(path).listFiles() ?: throw FileNotFoundException(
            "Directory has not been detected")

    for (file: File in listOfFiles) {
        // every note file from xpad with content of the note has 'content' word in the name
        if (!file.name.contains("content")) {
            continue
        }

        val content = file.readLines()
        if (content[0].contains("TODO")) { // first line need to contains 'TODO'
            return content.subList(1, content.size).toTypedArray()
        }
    }

    throw FileNotFoundException("No file with 'content' in name or word 'TODO' in the first line\n"
            + "has been detected in path:\n" + path)
}

private fun showNotifi(head: String, body: Array<String>, onlyInCommandLine: Boolean = false) {

    var lthLongestLine: Int = 0
    body.forEach { line: String ->
        if (lthLongestLine < line.length) {
            lthLongestLine = line.length
        }
    }

    // adding extra length if needed so notification window won`t be too short
    val miniLenght = config.minimalSpace
    if (lthLongestLine < miniLenght && !onlyInCommandLine) {

        val additionalSpace = " ".repeat(config.additionalSpace) //in space characters

        body[0] = body[0] + additionalSpace
    }

    var trueBody = ""
    for (line in body) {
        var copyOfLine = line // creates a copy so we can modify it

        val isGoodLine = (copyOfLine != "" || copyOfLine != "\n")
        if (!isGoodLine) { continue }

        if (line.startsWith("-")) { // we need to add space before '-' character
            copyOfLine = " " + copyOfLine // so notify-send won't threat this as a flag
        }

        if (trueBody == "") { // we need to
            trueBody = trueBody + copyOfLine + "\n"
            continue
        }

        trueBody = trueBody + "\n" + copyOfLine
    }

    if (onlyInCommandLine) {
        println(trueBody)
    } else {
        val icon = File(config.iconPath)
        val command = try {

            if (icon.exists()) {
                arrayOf("notify-send", "-i", icon.absolutePath, head, trueBody)
            } else {
                arrayOf("notify-send", head, trueBody)
            }

        } catch (e: Exception) {
            arrayOf("notify-send", "Error :(", "Pleas check config file, it is probably damaged or missing.")
        }

//        command.forEach { println(it) } // TESTING
        Runtime.getRuntime().exec(command)
    }
}

File: notificator.py
import os
import sys

# Constants
_USER_CONFIG_DIRECTORY_PATH = os.environ["HOME"] + "/.config"
_NOTIFICATOR_CONFIG_DIRECTORY_PATH = _USER_CONFIG_DIRECTORY_PATH + "/notificator"
_CONFIG_FILE_PATH = _NOTIFICATOR_CONFIG_DIRECTORY_PATH + "/config"

searched_phrase: str = ""
wants_help: bool = False
only_cmd: bool = False

help_msg = """This script is created to show content of xpad notes
                as notifications or command line messages.  
                   How to use:
                   xpad_notify <FLAGS> <PHRASE_FOR_SEARCH>
                   Flags:,
                       -c show notes only in command line
                       -h show help
                        Add phrase after flags to search in notes for it     
                   Config:
                      Go to .config/Notificator to edit config file.
                      You can set or change icon of the notification,
                      change path to where xpad is storing notes,
                      minimal length of notification 
                      and how much notification will be elongated."""


def main():
    _set_args()

    if wants_help:
        show_help()
    elif searched_phrase == "":
        show_notes()
    else:
        show_note_with()


def _set_args():
    global only_cmd, wants_help, searched_phrase
    args = sys.argv
    if len(args) > 1:
        first_arg = args[1]
        if first_arg[0] == "-":
            if "c" in first_arg:
                only_cmd = True
            if "h" in first_arg:
                wants_help = True

        else:
            searched_phrase = first_arg

        if len(args) > 2:
            searched_phrase = args[2]

    if len(sys.argv) > 2 and sys.argv[2] is not None:
        searched_phrase = sys.argv[2]


def show_note_with():
    notes = _get_all_notes()
    notes_with_phrase = []
    for note in notes:
        lines = open(note, 'r').readlines()
        for line in lines:
            if searched_phrase.lower() in line.lower():  # lower() function makes checking case insensitive
                notes_with_phrase.append(note)

    for note in notes_with_phrase:
        formatted_content = _format_content(_get_note_content(note))
        notify(formatted_content)


def _get_all_notes() -> list:
    all_notes = []

    with os.scandir(_USER_CONFIG_DIRECTORY_PATH + "/xpad") as entries:
        for entry in entries:
            if entry.is_file() and "content" in entry.name:
                all_notes.append(entry)
    return all_notes


def _get_note_content(note) -> list:
    return open(note).readlines()


def show_help():
    notify(help_msg)


def show_notes():
    notes = _get_all_notes()
    for note in notes:
        formatted_content = _format_content(_get_note_content(note))
        notify(formatted_content)


def _format_content(unformatted_content: list) -> str:
    form_content = ""
    for line in unformatted_content:
        if line == "\n" or line == " ":
            continue
        form_content = form_content + "\n" + line.strip()
    return form_content


def notify(content, header: str = "Notificator"):  # optional parameter cant be first
    if only_cmd:
        _print_notification(header, content)
    else:
        _show_notification(header, content)


def _show_notification(header: str, content: str):
    os.system(f"notify-send '{header}' '{content}'")


def _print_notification(header: str, content: str):
    os.system(f"echo '{header}' '{content}'")


if __name__ == '__main__':
    main()
